### ISSUE TITLE
[WebToolsE2E][Aspire][MacOS][Unstable]Debugging Aspire project using "dotnet watch", Safari browser page displays the error "Can't connect to the server" and manually refresh the browser can display the correct page. 

### ISSUE THREAD

**Issue by lemon0213** [dotnet/aspire#1008]
**REGRESSION INFO**: Not a Regression
**INSTALL STEPS**

1. Clean machine: MacOS Sonoma(14.1) on Intel machine
2. Install latest 8.0 SDK: 8.0.100(latest-released-build) 
3. Apply NuGet Feeds
4. Install the latest VS Code from: [Download Visual Studio Code - Mac, Linux, Windows](https://code.visualstudio.com/download)
5. Download the latest released C# Dev Kit from: [C# Dev Kit - Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit)
6. Open Terminal and run commands below to install aspire workload

- sudo dotnet workload update --skip-sign-check --interactive --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json

- sudo dotnet workload install aspire --skip-sign-check --interactive --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json

**REPRO STEPS**

1. Open VS Code, click Extension button > Install from VSIX… > select the downloaded "Green" file > Install.
2. Type 'Command+Shift+P' to open Command Palette > .NET: New Project > click .NET Aspire Application to create a new project
3. Right-click the project > Build
4. Right-click the aspire.AppHost project > Open in Integrated Terminal > dotnet watch.

**NOTE**: 
1. This issue may occur during the second or third time using dotnet watch to debug the project, I have tried many times, but the first debugging project rarely replicates this bug, so I added the unstable tag.
2. Using Edge browser displays the error "Can't reach this page", but the Edge browser can automatically refresh to the correct page.

**ACTUAL**
The Safari browser page displays the error "Can't connect to the server", then manually refresh the browser multiple times can display the correct page.

![image](https://github.com/dotnet/aspire/assets/66412274/e8ce9e22-906a-4367-b6bf-ea6e260b91b6)

**EXPECTED:**
The Safari browser can automatically refresh to the correct page.

**Comment by lemon0213** [comment]
Verify this issue using aspire 8.0 preview 1, the third time using dotnet watch to debug the project can meet the bug, added the REGRESSION INFO as Unintentional (bug).

**Comment by DamianEdwards** [comment]
This seems like it could be a timing issue in that the browser gets launched by `dotnet watch` but for some reason the endpoint the dashboard is on can't serve a response yet. Chrome and Edge will automatically retry connecting to the server if the first request fails to connect (e.g. DNS or TCP connection issue) but Safari might not.

@davidfowl could this be related to the decoupling of the dashboard from Aspire.Hosting? How is the "Now listening on.." message printed to the console coordinated to ensure it only happens once the server is running in the new design?

**Comment by smitpatel** [comment]
Currently dashboard starts after calls to start projects/containers. I think purpose of current way was to make sure things are starting to run when we show dashboard rather than empty dashboard. Perhaps dotnet watch could launch browser after "Now listening on.." appears.

**Comment by DamianEdwards** [comment]
> Perhaps dotnet watch could launch browser after "Now listening on.." appears.

I'm pretty sure that's what it does already as that logic is in `dotnet watch` itself (it sniffs console output for a known message).

**Comment by balachir** [comment]
@lemon0213 @v-cazhang this seems to be a 'dotnet watch' scenario on macOS. Does this issue still reproduce using Aspire 8.0 P2? How about when using latest Aspire 8.0 P3 builds?

**Comment by v-gaolisa** [comment]
@balachir Verified today, this issue does not repro on Aspire 8.0 P2, but it still repro using latest Aspire 8.0 P3 builds (8.0.0-preview.3.24072.7).

**Comment by danmoseley** [comment]
@kvenkatrajan should this be in P5, P6, or postponed?

**Comment by kvenkatrajan** [comment]
@adamint can you try reproing this in safari please?

**Comment by adamint** [comment]
@kvenkatrajan the issue (can't reach this page) appears in all browsers, and the specific safari behavior is reproducible

**Comment by adamint** [comment]
Blocked on above linked sdk issue

**Comment by v-polly** [comment]
@adamint , On Mac OS aspire preview 5 [8.0.0-preview.5.24201.12] + .NET 8.0.204 SDK(latest-released-build) with safari version 17.4.1 (19618.1.15.11.14), this issue is **not repro** with 'dotnet watch' / 'dotnet run -lp http/https'.
But it **repro** every time on safari with 'Ctrl + Shift + P' > Debug: Select and Start Debugging > C#: xxx.AppHost [https/http].
(Manually refresh the browser can display the correct page, Chrome and Edge not repro.)
![image](https://github.com/dotnet/aspire/assets/88425944/bd4e1ae1-7a64-4fda-816b-684334eac73b)

**Comment by joperezr** [comment]
@kvenkatrajan will take a look to see if there is something to fix for 8.1 and to check if this still repros.

**Comment by adamint** [comment]
This issue does reproduce with dotnet watch, all browsers. Also reproduces in VSCode with the above steps

**Comment by joperezr** [comment]
@kvenkatrajan is there anything to fix for 9.0 here?

**Comment by adamint** [comment]
> @kvenkatrajan is there anything to fix for 9.0 here?

There is not, the linked sdk issue is still active.

**Comment by davidfowl** [comment]
dotnet watch is being re-designed to work with aspire. That work is happening in 9.0.200 SDK which is not shipping with .NET 9. This has been fixed in that branch.

You can see the work items for watch and aspire here https://github.com/dotnet/sdk/issues?q=is%3Aissue%20label%3AAspire%20
